src/mono/mono/mini/mini.c

```diff
@@ -4481,3 +4481,56 @@ mini_get_cpu_features (MonoCompile* cfg)
 	// apply parameters passed via -mattr
 	return (features | mono_cpu_features_enabled) & ~mono_cpu_features_disabled;
 }
+
+int
+mini_primitive_type_size (MonoTypeEnum type)
+{
+	switch (type) {
+	case MONO_TYPE_I1:
+	case MONO_TYPE_U1:
+		return 1;
+	case MONO_TYPE_I2:
+	case MONO_TYPE_U2:
+		return 2;
+	case MONO_TYPE_I4:
+	case MONO_TYPE_U4:
+	case MONO_TYPE_R4:
+		return 4;
+	case MONO_TYPE_I8:
+	case MONO_TYPE_U8:
+	case MONO_TYPE_R8:
+		return 8;
+	case MONO_TYPE_I:
+	case MONO_TYPE_U:
+		return TARGET_SIZEOF_VOID_P == 8 ? 8 : 4;
+	default:
+		g_assert_not_reached ();
+		return 0;
+	}
+}
+
+/*
+ * mini_get_simd_type_info:
+ *
+ *   Return the element type of a SIMD type. Set NELEMS to the number of elements.
+ */
+MonoTypeEnum
+mini_get_simd_type_info (MonoClass *klass, guint32 *nelems)
+{
+	*nelems = 0;
+
+	const char *klass_name = m_class_get_name (klass);
+	if (!strcmp (klass_name, "Vector2") || !strcmp (klass_name, "Vector3") || !strcmp (klass_name, "Vector4") || !strcmp (klass_name, "Quaternion") || !strcmp (klass_name, "Plane")) {
+		*nelems = 4;
```

Comment:
I know this is just copying the existing code path, but the logic for Vector2/3 is currently unused and either needs a nice big comment explaining why `4` is correct for it or correspondingly needs to be `2/3`.

For Arm64 I don’t think there is any reason why `Vector2` shouldn’t be `2` since the platform natively has a `Vector64<float>`

For the `Vector3` and for `Vector2` on xarch, I could see `4` only being correct as part of some IR requirement in LLVM to ensure optimal codegen and that load/store boundaries would need to handle the correct transition from 2/3 to 4 elements

Reply:
Vector2 and Vector3 are currently not supported at all, thats why these didn't cause a problem so far.